ENH add possibility to have a callable for verbose_feature_names_out of ColumnTransformer

MarcBresson opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### What does this implement?

This brings the possibility to pass a callable to the `verbose_feature_names_out` parameter of `ColumnTransformer`. Instead of the new feature name being "transormer_name__feature_name", we could have "feature_name$this is amazing$TRANSFORMER_NAME".

#### Any other comments?

I have a few questions:

- How do you type sklearn as there are no stub files ?
- What is the version number I should put in `.. versionchanged` ?

In advance, thank you for your time.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

MarcBresson committed:
ENH add possibility to have a callable for verbose_feature_names_out of ColumnTransformer

adrinjalali commented:
cc @thomasjpfan 

pretty sure we had the conversation about this at some point.

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -655,11 +664,17 @@ def _add_prefix_for_feature_names_out(self, transformer_with_feature_names_out):
             names = list(
                 chain.from_iterable(
-                    (f"{name}__{i}" for i in feature_names_out)
+                    (feature_names_out_callable(name, i) for i in feature_names_out)
```
What do you think of allowing a string format to change instead of a callable?

```python
feature_names_out_template = "{transformer_name}${feature_name}"

chain.from_iterable(
    feature_names_out_template.format(
        transformer_name=transformer_name,
        feature_name=feature_name
    ) ...
)

```

BTW I am okay with a callable. I'm wondering if a string format is enough flexibility here.

MarcBresson replied:
I think this is a great idea to also allow for a string to format. It is a quicker way to do basic stuff, but it does not have the flexibility of a callable. For instance, it is impossible to put the transformer name in upper_case using a string to be formatted. Here is the doc about what is possible with string formatting: https://docs.python.org/fr/3.5/library/string.html#format-string-syntax. It shows limited possibilities.

thomasjpfan reviewed (commented):
Thank you for the PR!

MarcBresson committed:
ENH add possibility to provide a str to verbose_feature_names_out of ColumnTransformer

also: improve doc for the parameter, fix validation

MarcBresson committed:
TST: add tests for verbose_feature_names_out using callables or strings

MarcBresson commented:
Hey,

just added the possibility to use string format. I also added unit tests.

One of the test I created creates name clashing, and I remember seeing something about that somewhere in sklearn's doc. However, I can't seem to put my hand on it. From the behaviour of ColumnTransformer, I know it is not important since it uses np arrays rather than pandas dataframe, but do you think we should issue a warning somewhere in the doc?

thomasjpfan reviewed (commented):
> One of the test I created creates name clashing, and I remember seeing something about that somewhere in sklearn's doc. 

Name clashing should raise an error.

I'm not decided on supporting both callable and string. If the string format can support 90% of the use cases, then I'm okay with just having the string formatting.

adrinjalali commented:
The added complexity here for callable is small. I agree with @thomasjpfan on string formats probably enough, but since the added complexity is small, I don't mind either way.

MarcBresson commented:
what are the next steps for this PR?

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1652,3 +1676,13 @@ def _with_dtype_warning_enabled_set_to(warning_enabled, transformers):
     return result
+
+
+def _feature_names_out(
```
There are many `_feature_names_out` from the call-site. I think we should give this an more explicit name:

```suggestion
def _feature_names_out_with_str_format(
```

MarcBresson replied:
I'm not a fan. Creating new functions/methods actually hides functionalities away from the user. Indeed, they will often look for API reference for one particular function and check its parameters. However, I feel like it is rare that users will look at alternative functions.

In this case `_feature_names_out` is not bloated at all. The additional parameter makes it easy for the user to know that using string format is possible. I don't think it is an issue that there are many _feature_names_out from the call-site.

MarcBresson replied:
edit: nvm, it as been a long time since I've checked this PR. I forgot that `_feature_names_out` is a private function. I'll see what I can do :)

thomasjpfan commented on sklearn/compose/_column_transformer.py:
```diff
@@ -1652,3 +1676,13 @@ def _with_dtype_warning_enabled_set_to(warning_enabled, transformers):
+
+
+def _feature_names_out(
+    transformer_name: str, feature_name: str, str_format: str | None = None
```
Can we force the caller to always specific this?

```suggestion
    transformer_name: str, feature_name: str, str_format: str
```

thomasjpfan reviewed (commented):
Please add an entry to the change log at `doc/whats_new/v1.6.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

MarcBresson committed:
MAINT: rename _feature_names_out_with_str_format, make str_format parameter required

MarcBresson committed:
DOC: add PR 28934 to the changelog, add example of string format in verbose_feature_names_out description

MarcBresson commented:
Thank you! It is all done :)

thomasjpfan reviewed (dismissed):
<s>Minor comment, otherwise LGTM</s>

There is still a lint issue: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=68161&view=logs&j=32e2e1bb-a28f-5b18-6cfc-3f01273f5609&t=fc67071d-c3d4-58b8-d38e-cafc0d3c731a

MarcBresson committed:
DOC: follow contributing guideline for the typing in docstrings

MarcBresson commented on sklearn/compose/_column_transformer.py:
```diff
@@ -132,16 +133,30 @@ class ColumnTransformer(TransformerMixin, _BaseComposition):
         printed as it is completed.
 
-    verbose_feature_names_out : bool, default=True
+    verbose_feature_names_out : bool | str | Callable[[str, str], str], default=True
```
done :) sorry for the trouble

MarcBresson committed:
MAINT: sort imports for linting

adrinjalali commented on sklearn/compose/_column_transformer.py:
```diff
@@ -132,16 +133,29 @@ class ColumnTransformer(TransformerMixin, _BaseComposition):
         printed as it is completed.
 
-    verbose_feature_names_out : bool, default=True
+    verbose_feature_names_out : bool, str or Callable[[str, str], str], default=True
```
The rendered version of this now looks like:

![image](https://github.com/scikit-learn/scikit-learn/assets/1663864/c1fb29da-766f-42b2-ac07-1390dadafb78)

I think it makes sense to make a bullet point list for the 3 cases.

MarcBresson replied:
Ok it should be good. I'm gonna try to find how to compile the doc.

adrinjalali commented on sklearn/compose/_column_transformer.py:
```diff
@@ -132,16 +133,29 @@ class ColumnTransformer(TransformerMixin, _BaseComposition):
         If False, :meth:`ColumnTransformer.get_feature_names_out` will not
         prefix any feature names and will error if feature names are not
         unique.
+        If Callable[[str, str], str], :meth:`ColumnTransformer.get_feature_names_out`
```
```suggestion
        If ``Callable[[str, str], str]``, :meth:`ColumnTransformer.get_feature_names_out`
```

adrinjalali commented on sklearn/compose/_column_transformer.py:
```diff
@@ -132,16 +133,29 @@ class ColumnTransformer(TransformerMixin, _BaseComposition):
+        first argument of the callable is the transformer name and the
+        second argument is the feature name. The returned string will be the
+        new feature name.
+        If str, it must be a string ready for formatting. The given string will
+        be formatted using two field names: transformer_name and feature_name.
+        e.g. "{feature_name}__{transformer_name}".
+        See str.format method from the standard library for more info.
```
```suggestion
        If ``str``, it must be a string ready for formatting. The given string will
        be formatted using two field names: ``transformer_name`` and ``feature_name``.
        e.g. ``"{feature_name}__{transformer_name}"``.
        See :meth:`str.format` method from the standard library for more info.
```

adrinjalali reviewed (commented):
Otherwise LGTM.

MarcBresson committed:
DOC: use bullet points in docstrings and encapsulate python code samples in back quotes

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

MarcBresson commented:
@adrinjalali don't hesitate to review. I will rebase my branch right before merging so you can still see that everything passes.
